iroh-relay/src/dns.rs

```diff
@@ -1,22 +1,259 @@
-    sync::OnceLock,
+    fmt::Write,
+    future::Future,
+    net::{IpAddr, Ipv6Addr, SocketAddr},
+};
+
+use anyhow::{bail, Context, Result};
+use hickory_resolver::{lookup::TxtLookup, IntoName, ResolveError, Resolver, TokioResolver};
+use iroh_base::{NodeAddr, NodeId};
+use n0_future::{
+    time::{self, Duration},
+    StreamExt,
 };
+use url::Url;
 
-use anyhow::Result;
-use hickory_resolver::{Resolver, TokioResolver};
+pub mod node_info;
 
-/// The DNS resolver type used throughout `iroh`.
-pub(crate) type DnsResolver = TokioResolver;
+/// The n0 testing DNS node origin, for production.
+pub const N0_DNS_NODE_ORIGIN_PROD: &str = "dns.iroh.link";
+/// The n0 testing DNS node origin, for testing.
+pub const N0_DNS_NODE_ORIGIN_STAGING: &str = "staging-dns.iroh.link";
 
-static DNS_RESOLVER: OnceLock<TokioResolver> = OnceLock::new();
+/// The DNS resolver used throughout `iroh`.
+#[derive(Debug, Clone)]
+pub struct DnsResolver(TokioResolver);
 
-/// Get a reference to the default DNS resolver.
-///
-/// The default resolver can be cheaply cloned and is shared throughout the running process.
-/// It is configured to use the system's DNS configuration.
-pub fn default_resolver() -> &'static DnsResolver {
-    DNS_RESOLVER.get_or_init(|| create_default_resolver().expect("unable to create DNS resolver"))
+impl DnsResolver {
+    /// Create a new DNS resolver with sensible cross-platform defaults.
+    ///
+    /// We first try to read the system's resolver from `/etc/resolv.conf`.
+    /// This does not work at least on some Androids, therefore we fallback
+    /// to the default `ResolverConfig` which uses eg. to google's `8.8.8.8` or `8.8.4.4`.
+    pub fn new_with_defaults() -> Self {
+        let (system_config, mut options) =
+            hickory_resolver::system_conf::read_system_conf().unwrap_or_default();
+
+        // Copy all of the system config, but strip the bad windows nameservers.  Unfortunately
+        // there is no easy way to do this.
+        let mut config = hickory_resolver::config::ResolverConfig::new();
+        if let Some(name) = system_config.domain() {
+            config.set_domain(name.clone());
+        }
+        for name in system_config.search() {
+            config.add_search(name.clone());
+        }
+        for nameserver_cfg in system_config.name_servers() {
+            if !WINDOWS_BAD_SITE_LOCAL_DNS_SERVERS.contains(&nameserver_cfg.socket_addr.ip()) {
+                config.add_name_server(nameserver_cfg.clone());
+            }
+        }
+
+        // see [`DnsResolver::lookup_ipv4_ipv6`] for info on why we avoid `LookupIpStrategy::Ipv4AndIpv6`
+        options.ip_strategy = hickory_resolver::config::LookupIpStrategy::Ipv4thenIpv6;
+
+        let resolver = Resolver::tokio(config, options);
+        DnsResolver(resolver)
+    }
+
+    /// Create a new DNS resolver configured with a single UDP DNS nameserver.
+    pub fn new_with_single_nameserver(nameserver: SocketAddr) -> Self {
```

Comment:
IIRC the convention is `with_single_nameserver` without the `new` prefix.

Reply:
:+1: changed